distributed/diagnostics/plugin.py

```diff
@@ -282,34 +278,73 @@ def __init__(self, packages, pip_options=None, restart=False):
         self.packages = packages
         self.restart = restart
         self.pip_options = pip_options or []
+        self.id = f"pip-install-{uuid.uuid4()}"
 
     async def setup(self, worker):
-        from distributed.lock import Lock
-
-        async with Lock(socket.gethostname()):  # don't clobber one installation
-            logger.info("Pip installing the following packages: %s", self.packages)
-            proc = subprocess.Popen(
-                [sys.executable, "-m", "pip", "install"]
-                + self.pip_options
-                + self.packages,
-                stdout=subprocess.PIPE,
-                stderr=subprocess.PIPE,
-            )
-            stdout, stderr = proc.communicate()
-            returncode = proc.wait()
-
-            if returncode:
-                logger.error("Pip install failed with '%s'", stderr.decode().strip())
-                return
-
-            if self.restart and worker.nanny:
-                lines = stdout.strip().split(b"\n")
-                if not all(
-                    line.startswith(b"Requirement already satisfied") for line in lines
-                ):
-                    worker.loop.add_callback(
-                        worker.close_gracefully, restart=True
-                    )  # restart
+        from distributed.semaphore import Semaphore
+
+        async with (
+            await Semaphore(max_leases=1, name=socket.gethostname(), register=True)
+        ):
+            if not await self._is_installed(worker):
+                logger.info("Pip installing the following packages: %s", self.packages)
+                await self._set_installed(worker)
+                proc = subprocess.Popen(
```

Comment:
To avoid the `PipInstall` plugin from hogging the GIL if the installation of packages takes some time, we could switch to using asynchronous subprocesses here. One thing I'm a bit concerned about in that scenario is that we have no way of blocking a restart of the worker while the subprocess is running due to some other reason.

Reply:
Let's take one step at a time. This is only relevant on startup and during plugin initialization, the worker is not even connected to the scheduler, yet. I don't think anything will happen if we block here.

(Also, does Popen really block the GIL or just the event loop?)

Reply:
Sorry, I was not careful with the wording here. `subprocess.communicate()` hogs the _event loop_. After a small expedition down the GIL rabbithole, it looks like `subprocess.communicate()` does indeed release the GIL:
https://github.com/python/cpython/blob/e39ae6bef2c357a88e232dcab2e4b4c0f367544b/Modules/posixmodule.c#L8489-L8491

The visible effect remains the same, the worker might become unresponsive for a while, which will likely be exacerbated in #7103 since `conda install` doesn't feel like the fastest thing in the world.